### ISSUE TITLE
Review benchmarks

### ISSUE THREAD

**Issue by astrojuanlu** [astrojuanlu/sgp4-benchmarks#1]
Hi! @enritoomey @gerasdf @brandon-rhodes @bwinkel I created these benchmarks to compare four versions of SGP4 I found that are accessible from Python:

- https://github.com/brandon-rhodes/python-sgp4
- https://github.com/brandon-rhodes/python-sgp4/pull/76 (my Numba take)
- An open source fork we did in Satellogic to try NumPy vectorization before @brandon-rhodes released his lightweight C++ wrapper https://github.com/enritoomey/python-sgp4 (forked to https://github.com/astrojuanlu/python-sgp4/tree/numpy-vectorization/ to do some tweaks)
- https://github.com/bwinkel/cysgp4

These are the preliminary results from my computer (I ran these while on a Zoom meeting and with two PyCharm windows open, so by no means the CPU was "focused")

![Screenshot from 2020-12-10 17-20-34](https://user-images.githubusercontent.com/316517/101799360-5e240f80-3b0c-11eb-9df0-0c47e1464195.png)

<details>

Text:

```
------------------------------------------------------------------------------------------- benchmark 'test_multiple_satellites_multiple_dates': 4 tests ------------------------------------------------------------------------------------------
Name (time in ms)                                                                Min                 Max                Mean             StdDev              Median                 IQR            Outliers       OPS            Rounds  Iterations
---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
benchmarks/test_cpp_wrapper.py::test_multiple_satellites_multiple_dates       4.1670 (1.0)        8.0529 (1.11)       5.1026 (1.0)       0.8501 (1.14)       4.8190 (1.0)        1.1612 (1.75)         63;4  195.9798 (1.0)         213           1
benchmarks/test_numba.py::test_multiple_satellites_multiple_dates             5.2658 (1.26)       7.2534 (1.0)        5.9987 (1.18)      0.7461 (1.0)        5.8122 (1.21)       0.6628 (1.0)           1;0  166.7024 (0.85)          5           1
benchmarks/test_cython.py::test_multiple_satellites_multiple_dates            5.9947 (1.44)      64.1573 (8.85)      15.7319 (3.08)     11.9505 (16.02)     10.1389 (2.10)      14.5638 (21.97)        21;6   63.5653 (0.32)        124           1
benchmarks/test_pure_python.py::test_multiple_satellites_multiple_dates     481.0559 (115.44)   643.5991 (88.73)    570.9803 (111.90)   69.3276 (92.92)    594.0365 (123.27)   117.9784 (178.00)        2;0    1.7514 (0.01)          5           1
---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

----------------------------------------------------------------------------------------------------- benchmark 'test_single_satellite_multiple_dates': 5 tests ------------------------------------------------------------------------------------------------------
Name (time in us)                                                                    Min                    Max                  Mean                StdDev                Median                   IQR            Outliers          OPS            Rounds  Iterations
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
benchmarks/test_cpp_wrapper.py::test_single_satellite_multiple_dates             47.3840 (1.0)         233.4990 (4.01)        56.4084 (1.11)        13.0465 (2.95)        49.0530 (1.01)        12.4005 (2.67)    1783;1028  17,727.8520 (0.90)      12085           1
benchmarks/test_numba.py::test_single_satellite_multiple_dates                   47.8020 (1.01)         58.2840 (1.0)         50.6180 (1.0)          4.4261 (1.0)         48.6560 (1.0)          4.6455 (1.0)           1;0  19,755.8177 (1.0)           5           1
benchmarks/test_cython.py::test_single_satellite_multiple_dates                 120.7350 (2.55)     14,484.2320 (248.51)     623.6650 (12.32)    2,080.3012 (470.01)     138.3550 (2.84)        43.9160 (9.45)         5;21   1,603.4248 (0.08)        130           1
benchmarks/test_numpy_vectorized.py::test_single_satellite_multiple_dates       350.2610 (7.39)      1,486.9710 (25.51)      505.6298 (9.99)       194.8014 (44.01)      430.2280 (8.84)       182.9005 (39.37)      112;67   1,977.7314 (0.10)        823           1
benchmarks/test_pure_python.py::test_single_satellite_multiple_dates          3,747.0990 (79.08)    14,215.5790 (243.90)   6,135.4697 (121.21)   2,559.4007 (578.25)   5,004.0540 (102.85)   2,650.4435 (570.54)      36;20     162.9867 (0.01)        203           1
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

-------------------------------------------------------------------------------------------------------------- benchmark 'test_single_satellite_single_date': 5 tests -------------------------------------------------------------------------------------------------------------
Name (time in ns)                                                                   Min                       Max                    Mean                  StdDev                  Median                     IQR              Outliers  OPS (Kops/s)            Rounds  Iterations
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
benchmarks/test_cython.py::test_single_satellite_single_date                   138.9999 (1.0)         12,037.9991 (1.0)          195.5911 (1.0)          118.9539 (1.0)          159.0015 (1.0)           25.0002 (1.0)      9255;17164    5,112.7075 (1.0)      108366           1
benchmarks/test_cpp_wrapper.py::test_single_satellite_single_date              811.0001 (5.83)        19,229.9995 (1.60)       1,114.4280 (5.70)         436.1242 (3.67)         907.9995 (5.71)         429.9982 (17.20)    12919;5003      897.3213 (0.18)     122370           1
benchmarks/test_numba.py::test_single_satellite_single_date                  1,959.0007 (14.09)       32,922.9988 (2.73)      13,707.0005 (70.08)     15,672.1883 (131.75)     2,855.0021 (17.96)     27,632.5000 (>1000.0)         1;0       72.9554 (0.01)          5           1
benchmarks/test_pure_python.py::test_single_satellite_single_date           11,517.0005 (82.86)      130,464.9995 (10.84)     16,610.0263 (84.92)      6,611.0003 (55.58)     13,027.9986 (81.94)      8,165.9991 (326.64)     3459;842       60.2046 (0.01)      25651           1
benchmarks/test_numpy_vectorized.py::test_single_satellite_single_date     149,829.0003 (>1000.0)  1,940,334.9979 (161.18)   237,845.6489 (>1000.0)  113,301.2169 (952.48)   190,956.9983 (>1000.0)  123,047.0007 (>1000.0)     914;292        4.2044 (0.00)       5990           1
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
```
</details>

I wanted to reach out to you in case you notice something super wrong in how I'm using your respective libraries in the benchmarks. You can find them here:

https://github.com/astrojuanlu/sgp4-benchmarks/tree/master/benchmarks

I will publish the conclusions in the coming weeks as part of the work I'm doing for the OpenSatCom initiative that Libre Space Foundation is leading.

Also, of course if you have any suggestions I'll happily read your comments 😄

**Comment by bwinkel** [comment]
Hi @astrojuanlu,

that's awesome, thanks a lot! Very nice initiative.

I'm quite surprised by the results though. I'd naively assumed that cysgp4 would be bad in the single date / single time benchmark (at least in comparison), but have an advantage for the multi-benchmarks (given that this is completely encapsulated in Cython and powered by OpenMP). Can you say something more about the test system, i.e., number of CPU cores, architecture etc.? 

I'll try to run the benchmark on my machine later. If I understand it correctly, you're using pytest-benchmark? Could you share the necessary shell commands? (Or is running `pytest` sufficient, if all packages are present?)

I'm also quite interested in your OpenSatCom initiative. Working as a radio astronomer, I have ... strong feelings about the upcoming mega constellations 😁. Can you share some info?

**Comment by bwinkel** [comment]
> Could you share the necessary shell commands? (Or is running `pytest` sufficient, if all packages are present?)

Nevermind, I found the README ;-)

**Comment by bwinkel** [comment]
I looked at this in more detail. First, I ran the benchmark on a workstation, which is quite powerful in terms of CPU cores (24). This changes the results a bit towards cysgp4 - not surprising of course, as it is parallelized:

![few](https://user-images.githubusercontent.com/6582333/101834855-2b454000-3b3b-11eb-98e5-53d5301f665f.jpg)

However, it was still not too impressive, so I increased the problem space. For the single-sat multi-date test, I repeated the dates a 100 times. For the multi-sat multi-date test, I used 10000 TLEs instead of 100 (always the same, but for cysgp4 I can say, that it won't matter - not sure about the other packages). The results are very different then:

![many](https://user-images.githubusercontent.com/6582333/101834867-30a28a80-3b3b-11eb-9849-fc94370001a5.jpg)

It would probably worthwhile to play a little more with the element numbers to find out, what the sweet spot is (in terms of speed/core). There is certainly overhead with cython-openmp to spin-up the threads and the np.nditer.

By-the-way, I learned a lot from your test setup. It didn't occur to me that one could define the test data as a fixture to re-use it easily for all test functions. 👍

**Comment by astrojuanlu** [comment]
This is super helpful, thanks a lot!

> Can you say something more about the test system, i.e., number of CPU cores, architecture etc.?

Intel© Core™ i5-7200U CPU @ 2.50GHz × 2, Linux Mint 20 with 5.4.0-56-generic kernel. I will be careful to run the benchmarks in a dedicated machine before publication.

> It would probably worthwhile to play a little more with the element numbers to find out, what the sweet spot is (in terms of speed/core). There is certainly overhead with cython-openmp to spin-up the threads and the np.nditer.

Absolutely! I will add a few more cases that includes more data points as you said.

> By-the-way, I learned a lot from your test setup. It didn't occur to me that one could define the test data as a fixture to re-use it easily for all test functions. :+1:

🙌🏽 

> I'm also quite interested in your OpenSatCom initiative. [...] Can you share some info?

We will publish more info on https://opensatcom.org/ soon, and also we'll mention it a bit in the upcoming https://events.libre.space/event/4/ _this weekend_ (you have a few minutes left to register! in any case, the talks will be recorded and published)

> Working as a radio astronomer, I have ... strong feelings about the upcoming mega constellations 😁 

Yeah, I can understand...! It's a contested topic, and ~Elon fanboys~ excited Twitter crowds make it even more controversial. The only thing I can say is that I'm concerned as well, and that (as far as I understand) this project is focused on looking for open source solutions for the SatCom industry. @ppapadeas might want to add something :)

**Comment by bwinkel** [comment]
> We will publish more info on https://opensatcom.org/ soon, and also we'll mention it a bit in the upcoming https://events.libre.space/event/4/ _this weekend_ (you have a few minutes left to register! in any case, the talks will be recorded and published)

Unfortunately, I won't be able to make it this weekend, but I'm looking forward to watch the recordings. Thanks.

**Comment by brandon-rhodes** [comment]
@astrojuanlu — Thanks for the heads up!

The major problem with the benchmark with respect to my library, `sgp4`, is that `test_single_satellite_single_date()` is asking it to do more work than its main competitor, the Cython-powered library. The benchmarked call to `satrec.sgp4()` both calls the underlying C++ routine by Vallado, and then bundles up the result as Python floating point objects inside of tuples — a very expensive operation that requires multiple mallocs. But the benchmarked call to the Cython `sat.eci_pos()` only computes the vectors internally, without producing the tuples and floats that return the result to the user.

A more nearly equivalent comparison would be to ask the Cython version to malloc the tuples and floats as well:

```
    def f():
        p = sat.eci_pos()
        return p.loc, p.vel

    r, v = benchmark(f)
```

This brings the libraries nearly even in performance on my laptop, though I must admit that Cython still has the edge! I wonder if the Cython is somehow more amenable to optimization — or if my `setup.py` is missing options that would compile Vallado’s code more tightly than is currently the case?

In any event, I'm very happy to see the work you've done to put this comparison together, and I hope that your audience finds the numbers useful. I'll let you know if I run across any improvements I can make to `sgp4` in the one-satellite-one-time case where it still seems to lag a bit!

**Comment by bwinkel** [comment]
> This brings the libraries nearly even in performance on my laptop, though I must admit that Cython still has the edge! I wonder if the Cython is somehow more amenable to optimization — or if my `setup.py` is missing options that would compile Vallado’s code more tightly than is currently the case?

One should not forget, that cysgp4 is not a wrapper around the Vallado software, but uses a different implementation, by D. Warner. One would have to do C/C++ benchmarks to assess the original speed ratio between the two, but that's beyond what I could currently invest (time-wise) into this...

**Comment by brandon-rhodes** [comment]
> uses a different implementation, by D. Warner

Oh! I hadn't realized, thanks for the clarification. It could, then, simply be a faster means of propagating.

**Comment by astrojuanlu** [comment]
I think the trick is in the parallelization that @bwinkel mentioned, I forgot about it even after thoroughly reading your conversation on https://github.com/bwinkel/cysgp4/issues/8 

Which, by the way, reminds me that Numba has parallelization support https://numba.readthedocs.io/en/stable/user/parallel.html I will give this a try. Isn't benchmarking fun? 😈

**Comment by brandon-rhodes** [comment]
> I think the trick is in the parallelization that @bwinkel mentioned…

I'm not sure if you're responding to my comments or some of the others, but my focus was on the `test_single_satellite_single_date` case, which I don't think would involve any parallelization, but is probably simply a metric of how fast the single-case computation runs.

**Comment by astrojuanlu** [comment]
Oh you're right, I mixed up things here (too much multi tasking)

**Comment by astrojuanlu** [comment]
I have pushed several changes to `master` based on your feedback, namely:

- Wrapped all "single satellite, single date" benchmarks in a short function so we are measuring the same things (thanks @brandon-rhodes for the suggestion!)
- Renamed "multiple dates" cases to "medium" and added "large" cases with 100x more TLEs and 100x more dates (thanks @bwinkel!)

In parallel (no pun intended), I also made some changes in https://github.com/brandon-rhodes/python-sgp4/pull/76 to leverage `@njit(parallel=True)` capabilities, with promising results.

I consider this "first pass" over the benchmarks complete. There are of course many more things we could do: plot runtime vs number of satellites / dates, measure RAM consumption... But I have some deadlines I want to meet, so I will stop here, run these both in my laptop and in a powerful dedicated machine, write a report with my findings, and publish it. The code will remain open source as it is now, so of course everybody is welcome to contribute suggestions and pull requests.

Thanks again for your prompt review, closing!
